features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c

```diff
@@ -11,31 +11,59 @@
 
 #include <string.h>
 #include "cmsis_os2.h"
+#include "rtx_os.h"
+#include "stdlib.h"
 
 /* This is an example OS abstraction layer for CMSIS-RTOSv2 */
 
 void *os_wrapper_thread_new(const char *name, int32_t stack_size,
                             os_wrapper_thread_func func, void *arg,
                             uint32_t priority)
 {
-    osThreadAttr_t task_attribs = {.tz_module = 1};
+    osRtxThread_t *thread;
+    void *stack;
 
-    /* By default, the thread starts as osThreadDetached */
-    if (stack_size != OS_WRAPPER_DEFAULT_STACK_SIZE) {
-        task_attribs.stack_size = stack_size;
+    thread = (osRtxThread_t *)malloc(osRtxThreadCbSize);
+    if(thread == NULL){
+        return NULL;
     }
-    task_attribs.name = name;
-    task_attribs.priority = (osPriority_t) priority;
 
-    return (void *)osThreadNew(func, arg, &task_attribs);
+    stack = (void *)malloc(stack_size);
+    if(stack == NULL){
+        free(thread);
+        return NULL;
+    }
+
+    const osThreadAttr_t thread_attribs = {
+#if defined (__ARM_FEATURE_CMSE) &&  (__ARM_FEATURE_CMSE == 3U)
+        .tz_module = 1,
+#endif
+        .name = name,
+        .priority = (osPriority_t) priority,
+        .cb_mem = thread,
+        .cb_size = osRtxThreadCbSize,
+        .stack_size = stack_size,
+        .stack_mem = stack};
```

Comment:
Could we use Mbed OS C++ Thread class instead? Then we don't need to duplicate the know-how of how to create an Mbed OS thread here. For instance, if the Mbed OS C++ Thread class decided to change if threads were "joinable", we wouldn't pick up that change here, even though in current implementation the behavior is aligned. This wrapper code would be more robust if we'd avoid repeating ourselves (DRY) and use the C++ Thread class.

Reply:
Mbed OS C++ thread class always sets the thread to be detached state by default and have not seen any occurrences of joinable state using mbed-os C++ class. I don't see a reasonable argument to make so many changes (almost rewriting the wrapper) to increase the robustness at the moment.

Reply:
Joinable state was just one example. Any part of this could change. For instance, the memory layout of mutexes in RTX could change, requiring updates to both here and the Mutex class, rather than just the Mutex class.

Maintainability is an extremely strong reason to block adding this code in its current shape.

Reply:
I think we need to align with TF-M, as these wrappers are only to be used with them. If Mbed OS C++ decides to make it as "joinable" and if TF-M API's think them to be "Detachable" thread then it might be a problem.

If my understanding is correct, For example, the regression tests don't use `thread_join()` API, they just support mostly "detachable" state operations.

Reply:
Anything could change, not just joinability. For instance, Mbed OS may want to change how stacks are allocated. Having one place for the logic to allocate stacks is best for maintainability. Let's re-use the Mbed OS code we have rather than reimplementing it again in C and just for TF-M; it's better for maintainability and also for code size (since we'll have less code).